### PR TITLE
Burgle.lic add pawn and bin support

### PR DESCRIPTION
**Pull request by vtcifer**
**Changes**:

*burgle.lic:*
- Added pawn and bin support
- Added new command line argument for overriding home town for pawn/bin
- Added another match for sneaking failures

*base-town.yaml:*
- Added new pawnshops

*common-travel.lic:*
- Added new generic routine to get roomid of targets from base-town: get_hometown_target_id(hometown, target)

*common.lic*
- Added array of all supported hometowns.
- Fixed regex match for 'fix_dr_bullshit'.  It broke when items had punctuation in their adj/noun.
- Added left_hand_noun/right_hand_noun functions.

### CODE REVIEW

File: burgle.lic [hunk 1]
```diff
@@ -118,9 +125,47 @@ class Burgle
...
+    else
+      message("You are not a thief.  You can't use thief bins.  Attempting fallback to pawning loot.")
+      pause 5
+      @loot_type = 'pawn'
+      validate_pawn_settings
+    end
+  end
+
+  def validate_pawn_settings
+    # pawn only works if there is a pawnsho pin your hometown
```
Comment (rpherbig) [inline burgle.lic]: 💡 'pawn shop in'

File: burgle.lic [hunk 2]
```diff
@@ -118,9 +125,47 @@ class Burgle
...
+    # fall abck to drop if you can't pawn, next best option since you don't want to keep the loot
```
Comment (rpherbig) [inline burgle.lic]: 💡 'fall back to'

File: burgle.lic [hunk 3]
```diff
@@ -118,9 +125,47 @@ class Burgle
...
+      validate_pawn_settings
+    else
+      @loot_room_id = nil
+    end
+
     return true
   end
 
+  def validate_bin_settings
+    # bin only works if your a thief, and there is a bin in your hometown
```
Comment (rpherbig) [inline burgle.lic]: 💡 'your' -> 'you're'

File: burgle.lic [hunk 4]
```diff
@@ -20,12 +20,18 @@ class Burgle
         { name: 'start', regex: /start/,optional: false, description: 'Required: prevents accidentally running burgle and protects against burgle typo for starting script' },
         { name: 'entry', options: %w[lockpick rope cycle], optional: true, description: 'Override yaml setting for entry_type.' },
         { name: 'roomid', regex: /\d+/, optional: true, description: 'Override yaml setting and go to room id (#) specified.  The room shold be chosen VERY carefully.'},
-        { name: 'loot_type', options: %w[drop keep], optional: true, description: 'Override yaml setting for loot. (items on item_whitelist are always kept.)' },
+        { name: 'loot_type', options: %w[drop keep pawn bin], optional: true, description: 'Override yaml setting for loot. (items on item_whitelist are always kept.)' },
+        { name: 'hometown', options: $HOMETOWN_LIST, optional: true, description: 'Overrid yaml hometown settings for bin and pawn.  If no bin or pawnshop in that hometown, loot_type will revert to drop.'}
```
Comment (rpherbig) [inline burgle.lic]: 💡 'Override'

File: burgle.lic [hunk 5]
```diff
@@ -205,26 +250,9 @@ class Burgle
...
-        when /^What were you referring to/  #handle items that stack/missing items
-          next
-        end
-      end
-    #when 'pawn' -> not implemented yet
-      #goto closest pawnshop, go through bags and get items to drop, unless on whitelist
-    #when 'bin'  -> not implemented yet
-      #check for thief, if so, goto bin, go through bags and get items to bin, unless on whitelist
-    end
+    process_loot(@loot_type)
```
Comment (rpherbig) [inline burgle.lic]: ❔ Why pass in `@loot_type` when it's accessible from within the function?

File: common-travel.lic [hunk 6]
```diff
@@ -180,4 +180,24 @@ module DRCT
...
+    if !target_id
+      echo("*** get_hometown_target_id failed first attmpt for #{target} in #{hometown}. Trying again:") if $common_travel_debug
+      pause 2
+      hometown_data = get_data('town')[hometown]
+      target_id = hometown_data[target] && hometown_data[target]['id']
+      if !target_id
+        echo("*** get_hometown_target_id failed second attempt for #{target} in #{hometown}.  Likely target doesn't exist") if $common_travel_debug
+        target_id = nil
+      else
+        echo("*** get_hometown_target_id succeded second attempt for #{target} in #{hometown}.") if $common_travel_debug
```
Comment (rpherbig) [inline common-travel.lic]: 💡 'succedeed'

File: common-travel.lic [hunk 7]
```diff
@@ -180,4 +180,24 @@ module DRCT
     end
     return reverse_path
   end
+
+  def get_hometown_target_id(hometown, target)
+    hometown_data = get_data('town')[hometown]
+    target_id = hometown_data[target] && hometown_data[target]['id']
+    #try twice with pause if failed the first time, because sometimes data doesn't get loaded correctly
+    if !target_id
+      echo("*** get_hometown_target_id failed first attmpt for #{target} in #{hometown}. Trying again:") if $common_travel_debug
```
Comment (rpherbig) [inline common-travel.lic]: 💡 'attempt'

File: common-travel.lic [hunk 8]
```diff
@@ -180,4 +180,24 @@ module DRCT
     end
     return reverse_path
   end
+
+  def get_hometown_target_id(hometown, target)
+    hometown_data = get_data('town')[hometown]
+    target_id = hometown_data[target] && hometown_data[target]['id']
+    #try twice with pause if failed the first time, because sometimes data doesn't get loaded correctly
```
Comment (rpherbig) [inline common-travel.lic]: ❓ I've never seen data fail to load. Can you be more specific with what's going on?

Comment (vtcifer) [reply common-travel.lic]: For whatever reason, every so often the roomid data for binning in steal.lic doesn't get loaded:
```
> ,steal
--- Lich: steal active.
--- Lich: jail-buddy active.
[steal: *** Binning is not yet supported in Shard]
```
Binning is definetly supported in Shard.  I also get the same thing for Crossing as well.

The only thing that triggers that message (in steal) is when you're set to bin, and the below code fails. 

https://github.com/rpherbig/dr-scripts/blob/f225d0fbaf81d903ee2ad71f56831fd7064c0718/steal.lic#L24-L30

Others have reported the same intermittently in lnet, to which advice has generally been, if you notice it, kill it restart, and it'll pick up.  I added a 2nd check to my steal.lic and never had problems past the 2nd load attempt (never got around to doing a PR).

I ran into the same issue when introducing this generic function, unless I tried a second time if it failed the first.

**Pull request merged**